### PR TITLE
add a README in src/backend/fts/

### PR DESCRIPTION
**Pull request by cng001**
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [X ] Document changes
- [] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@cng001 Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=8196)!

[Click here](https://cla.pivotal.io/sync/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=8196) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@cng001 Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=8196)!

File: src/backend/fts/README [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+   node) only. It is started by ServerLoop() of postmaster. There we start
+   a group of GP background processes through a structure PMSubProcList.
+   PMSubProcList is an array and each entry contains the description of a
+   GP background process, including the main entry function of the process
+   and a flag to indicate on what type (master, segment, or both) of nodes
+   this process can be started. The first entry in the array is about the
+   FTS probe. When its flag is checked, it says only  Gp_entry_postmaster
+   can start it.
+
+   PR 7364 has made some changes to the above implementation recently and
```
Comment (ashwinstar) [inline src/backend/fts/README]: We document the current state of the code and leave the historical context to be provided by git. So, would like if can take out PR reference and only mention how the code behaves currently in master throughout this README.

File: src/backend/fts/README [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        Gp_entry_postmaster && Gp_role == GP_ROLE_DISPATCH
+
+   and start the FTS probe process only when the condition is true. All
+   these are done in load_auxiliary_libraries() which now is responsible
+   for starting all GP background processes.
+
+   Note: load_auxiliary_libraries() is called outside of ServerLoop().
+         Not sure about the implication of this difference.
+
+2. The FTS probe process usually runs in a infinite loop. A few things
```
Comment (ashwinstar) [inline src/backend/fts/README]: s/usually/always?

File: src/backend/fts/README (hunk 1 above)
Comment (cng001) [reply src/backend/fts/README]: done

File: src/backend/fts/README (hunk 2 above)
Comment (cng001) [reply src/backend/fts/README]: changed

File: src/backend/fts/README [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+   called first to make sure this dialog is for FTS requests and thus the
+   Postgres process spawn for it would be a FTS handler(am_ftshandler = true).
+   Then it accepts the request and process the ‘Q’ type message using
+   HandleFtsMessage(). This function deals with three kinds of requests:
+
+        - Probe
+        - Sync
+        - Promote
+
+6. Not sure if FTS can be stopped by SIGUSR2 now.
```
Comment (pengzhout) [inline src/backend/fts/README]: SIGUSR2 is ignored by FTS now, like other background, postmaster use SIGTERM to stop the FTS.

Comment (cng001) [reply src/backend/fts/README]: replace point 6 with "SIGUSR2 is ignored by FTS now, like other background, postmaster use SIGTERM to stop the FTS."

File: src/backend/fts/README [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+   is down and postmaster needs to have a chance to bring it up again.
+
+   This is how the FTS probe process is started.
+
+
+2. The FTS probe process always runs in a infinite loop and trying to
+   do one round of polling at each iteration to get the health status
+   of all segments. At each iteration, it calls WaitLatch with timeout
+   to block itself for a while. Thus, two types of events might actually
+   trigger the polling actually. One is timeout on the latch, and the
```
[the remark this answers is not in the document]
Comment (cng001) [reply src/backend/fts/README]: removed

File: src/backend/fts/README [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+   In many code running on the master node, e.g. inside the code of the
+   function cdbgang_createGang_async(), if we hit the PG_CATCH block,
+   FtsNotifyProber() is called to signal postmaster with SIGINT, and
+   the postmaster’s SIGINT handler calls SetLatch() to unblock the FTS
+   main loop if it is now waiting on the Latch.
+
+   This is the way how the FTS probe process starts a round of polling.
+
+3. On the master node side, the FTS probe process gets the configuration
+   from catelog table pg_segement_configure, which describes the status of
```
[the remark this answers is not in the document]
Comment (cng001) [reply src/backend/fts/README]: fixed

File: src/backend/fts/README [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+   each segment and also tells if any of them has a mirror. In a simple
+   case, we can think of that, for each unique content(or segindex) value,
+   we may see a primary segment and a mirror segment. The two make a pair
+   and they have the same content(or segindex) value but different dbid.
+   The FTS’s job is to periodically (sometimes event driven) poll all the
+   segments that have mirror setup. If any of them is in down status, FTS
```
[the remark this answers is not in the document]
Comment (cng001) [reply src/backend/fts/README]: The point 2 is changed as below: 
" On the master node side, the FTS probe process gets the configuration
   from catelog table gp_segment_configuration, which describes the status
   of each segment and also tells if any of them has a mirror. In a simple
   case, we can think of that, for each unique content(or segindex) value,
   we may see a primary segment and a mirror segment. The two make a pair
   and they have the same content(or segindex) value but different dbid.
   
   FTS probes only the primary segments. Primary segments provide their own 
   status as well as their mirror's status in response. When a primary segment 
   is found to be down, FTS promotes its mirror, only if it was in-sync with 
   the primary. If the mirror is out-of-sync, this is considered "double 
   failure" and FTS does nothing. The cluster is unusable in this case.
   
   In short, the FTS probe gets the needed information of all segments 
   from these catalog tables and also update them according to the dynamically 
   changing runtime status.
"

File: src/backend/fts/README [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+
+2. The FTS probe process always runs in a infinite loop and trying to
+   do one round of polling at each iteration to get the health status
+   of all segments. At each iteration, it calls WaitLatch with timeout
+   to block itself for a while. Thus, two types of events might actually
+   trigger the polling actually. One is timeout on the latch, and the
+   other one is that someone sets the latch.
+
+   In many code running on the master node, e.g. inside the code of the
```
[the remark this answers is not in the document]
Comment (cng001) [reply src/backend/fts/README]: I am trying to avoid the specific code as much as possible but still prefer some clue to give a clue to developers a minimum clue where the logic is implemented and go there quickly. So I changed the point 1 and 2 as below: 
" 
1. The FTS probe process is running on the Gp_entry_postmaster (master
   node) only. It starts as a background worker process managed
   by the BackgroundWorker structure. (see src/include/postmaster/bgworker.h).
   Greenplum sets up a group of GP background processes through an
   array structure PMAuxProcList. There each entry is for one GP background
   process.

   The most important members of the BackgroundWorker structure are two
   function pointers. One is the pointer of the main entry function of the
   GP background process and the other one is the pointer of the function
   that determine if the process should be started. For FTS, the two
   pointers are FtsProbeMain() and FtsProbeStartRule(), respectively.
   This is hard-coded in postmaster.c.

         #define MaxPMAuxProc 6 
         static BackgroundWorker PMAuxProcList[MaxPMAuxProc]   ...

     In Postmaster, we will check the following condition:

        Gp_entry_postmaster && Gp_role == GP_ROLE_DISPATCH

   The FTS probe process is started when the condition is true.

   In the initialization phase, we register one BackgroundWorker entry
   for each GP background process into postmaster's private structure
   BackgroundWorkerList. When we do this, the above condition is checked
   to decide if FTS should be registered there or not. The reader may
   want to check load_auxiliary_libraries() for more detial.

   Later, the postmaster tries to start the processes that have been
   registered in the BackgroundWorkerList, which includes the FTS
   probe process. For the cases that the first attempt to start
   a particular process fails, or a process goes down for some reason
   and then needs to be brought up again, postmaster gets the chances
   to re-start it in its main loop. At a iteration, it checks
   the status of these processes and acts accordingly.

   This is how the FTS probe process is started.


2. The FTS probe process always runs in a infinite loop and trying to
   do one round of polling at each iteration to get the health status
   of all segments. At each iteration, it wait one a latch with timeout
   to block itself for a while. Thus, two types of events might
   trigger the polling. One is timeout on the latch it is waiting for,
   and the other one is that someone sets the latch.

   Certain components running on master node may interrupt FTS from its
   wait to trigger a probe immediately. This is referred to as notifying
   FTS. Dispatcher is one such component. As an example, it can notify
   FTS if it encounters an error while creating a gang. The reader may
   check FtsNotifyProber() to find more cases.

   This is the way how the FTS probe process starts a round of polling.
"

File: src/backend/fts/README [hunk 8]
```diff
@@ -35,12 +35,18 @@ Tolerance Service (FTS):
...
         Gp_entry_postmaster && Gp_role == GP_ROLE_DISPATCH
 
-   and start the FTS probe process only when the condition is true. All
-   these are done in load_auxiliary_libraries() which now is responsible
-   for starting all GP background processes.
+   and start the FTS probe process only when the condition is true.
+
+   The load_auxiliary_libraries() function called in PostmasterMain()
+   will add a worker entry into postmaster's private structure
+   BackgroundWorkerList. Then PostmasterMain() will call
```
Comment (pengzhout) [inline src/backend/fts/README]: maybe_start_bgworker() is called in PostmasterMain() to try to start a background worker whose
bgw_start_time be set to BgWorkerStart_PostmasterStart (which means can be started even the startup process haven't finished the startup of xlog). So actually, FTS process is started inside ServerLoop() by maybe_start_bgworker().

Comment (pengzhout) [reply src/backend/fts/README]: It's outdated, please ignore my comment.

**Comment by asimrp** [comment]
Thank you @cng001 for incorporating all the comments.  The readme is looking great now.  I found the background worker and startup sequence details particularly helpful.  Appreciate your contribution very much!

**Comment by danielgustafsson** [comment]
> On 24 Jul 2019, at 10:05, Asim R P <notifications@github.com> wrote:
> 
> Appreciate your contribution very much!

+1, this is great.

**Comment by ashwinstar** [comment]
I am looking into this, will do some final touch-ups and then will get this committed today. Thanks a lot.

**Comment by ashwinstar** [comment]
I committed this PR https://github.com/greenplum-db/gpdb/commit/45558a78b235e3c8060403335868de68caf6fc28. Thanks again @cng001, keep them coming.
